File: src/day14.cpp
#include <algorithm>
#include <regex>
#include <string>
#include <vector>
#include "aocutil.h"

const std::string pattern{R"((\w+) can fly (\d+) km/s for (\d+).* must rest for (\d+) seconds.)"};
const std::regex re{pattern};
const int Target_Time = 2503;

struct Rule {
    int Speed, Duration, Rest;
};

int get_distance(int, Rule);

aoc_result_t day14()
{
    std::vector<Rule> rules{};
    get_input_lines(14, false, [&](std::string line) {
        std::smatch sm;
        std::regex_match(line, sm, re);
        rules.push_back(Rule{std::stoi(sm[2]), std::stoi(sm[3]), std::stoi(sm[4])});
    });
    int rsz = rules.size();

    // Part 1
    int part1 = 0;
    for (auto& rule : rules) {
        int dist = get_distance(Target_Time, rule);
        if (dist > part1) {
            part1 = dist;
        }
    }

    // Part 2
    std::vector<int> points(rsz, 0);
    for (int t = 1; t <= Target_Time; t++) {
        int furthest = 0;
        int furthestIndex = 0;
        for (int r = 0; r < rsz; r++) {
            Rule rule = rules[r];
            int dist = get_distance(t, rule);
            if (dist > furthest) {
                furthest = dist;
                furthestIndex = r;
            }
        }
        points[furthestIndex] += 1;
    }
    int part2 = *std::max_element(points.begin(), points.end());

    return {part1, part2};
}

int get_distance(int time, Rule rule)
{
    int mod = rule.Duration + rule.Rest;
    auto dv = std::div(time, mod);
    int result = dv.quot * rule.Speed * rule.Duration;
    if (dv.rem >= rule.Duration) {
        result += rule.Speed * rule.Duration;
    } else {
        result += rule.Speed * dv.rem;
    }
    return result;
}

File: src/day14.py
import re

Target_Time = 2503

def get_distance(time, speed, duration, rest):
    mod = duration + rest
    tt = time // mod
    result = speed * duration * tt
    rr = time % mod    
    if rr > duration:
        result += speed * duration
    else:
        result += speed * rr
    return result

f = open("../input/day14-input.txt", "r")
input = f.readlines()

rr = re.compile(r"(\w+) can fly (\d+) km/s for (\d+).* must rest for (\d+) seconds.")
rules = [mm[0] for mm in [rr.findall(line) for line in input]] 

# Part 1
dists = [get_distance(Target_Time, int(r[1]), int(r[2]), int(r[3])) for r in rules]
print("Part 1: ", max(dists)) 

# Part 2
stats = [0 for line in input]
t = 0
while t < Target_Time:
    t += 1
    furthest = 0
    furthestIndex = -1
    i = 0
    for r in rules:
        dist = get_distance(t, int(r[1]), int(r[2]), int(r[3]))
        if dist > furthest:
            furthest = dist 
            furthestIndex = i
        i += 1
    stats[furthestIndex] += 1

print("Part 2: ", max(stats))

